### PR TITLE
fix(plan): drive fulltext index from an aggregate MATCH proven present by HAVING (#29065)

### PR DESCRIPTION
**Pull request by cpegeric**
## What type of PR is this?

- [ ] API-change
- [x] BUG
- [ ] Improvement
- [ ] Documentation
- [ ] Feature
- [ ] Test and CI
- [ ] Code Refactoring

## Which issue(s) this PR fixes:

issue #29065

## What this PR does / why we need it:

A grouped query whose only MATCH is an aggregate -- `MAX(match) AS score ... HAVING score > 0` or `HAVING MAX(match) > 0` -- had no scan-level driver, so the aggregate MATCH was only a consumer: `served` stayed empty and the raw fulltext_match reached execution as error 20105.

getFullTextMatchFromAggHaving collects the match inside a MAX/SUM aggregate that a membership-implying HAVING (`agg(match) > c` / `>= c`, a relevance of 0 failing it) proves must be present, and feeds it as a driver; the existing #28681 rewrite then turns the AggList arg into the served score and the HAVING colref resolves to it.

Safety gates, because driving INNER-joins the scan to matchers BEFORE aggregation (dropping each group's non-matching rows):
  - safeAggForFullTextDriver: MAX/SUM only. MIN(match)>0 means every row matches, and AVG/COUNT change value once non-matchers are dropped.
  - aggOutputInvariantToMatcherFilter: drive ONLY when every AggList entry is a MAX/SUM of a driven match and no grouping key is a match. Any co-aggregate -- COUNT(*), AVG, or an aggregate over another column/different match -- would be silently computed over matchers only for a multi-row group, so those are left at 20105.
  - membership operators only (`>`/`>=` with the WHERE-side nonNegativeConstValue rule); unwrapMonotoneScalar sees through the cast/round the comparison inserts.

Tests: TestFullTextAggHavingDrivesIndex (driven) and TestFullTextAggHavingCoAggregateNotDriven (co-aggregate refused). BVT fulltext_agg_having / fulltext2_agg_having cover alias and direct MAX, SUM, EXPLAIN(index scan), and the MIN / co-aggregate / COUNT safety cases pinned at 20105.

### CODE REVIEW

**Review by XuPeng-SH — CHANGES_REQUESTED** [review]
### Deep review（head aeef9dab7a5c916aacb7fdcc6665de219a03c056）

结论：REQUEST_CHANGES。发现 2 个 P1 静默错结果问题，未确认 P0。

1. [P1] 多个 aggregate MATCH 被错误转换为文档级交集。见 [apply_indices_fulltext.go:2434-2438](https://github.com/matrixorigin/matrixone/blob/aeef9dab7a5c916aacb7fdcc6665de219a03c056/pkg/sql/plan/apply_indices_fulltext.go#L2434)。当前只验证每个 MAX/SUM 对应某个 MATCH，但下游会把所有 MATCH 流按 doc_id INNER JOIN，要求同一文档同时匹配所有条件。若同一分组内不同文档分别满足 alpha/beta，聚合组会被提前删掉，结果错误；FULLTEXT/FULLTEXT2 共用该路径。

2. [P1] 跨 WINDOW 收集外层 FILTER，可能把窗口后的过滤错误当成 HAVING。见 [apply_indices.go:858-870](https://github.com/matrixorigin/matrixone/blob/aeef9dab7a5c916aacb7fdcc6665de219a03c056/pkg/sql/plan/apply_indices.go#L858)。路径搜索任意 FILTER 并穿过单输入节点寻找 AGG，没有验证 WINDOW 语义边界。外层按 aggregate score 过滤时，提前只保留匹配组会改变 ROW_NUMBER 等窗口结果。

建议仅使用真正属于该 AGG 的 HAVING，拒绝不满足所有 MATCH 不变性证明的多 MATCH 联合驱动，并补多行分组、多 MATCH、窗口后过滤的 FULLTEXT/FULLTEXT2 结果 BVT。

其余 HAVING 识别、GROUP BY/NULL、join/outer join、index eligibility、资源和错误路径未发现新的确定性问题。当前 head 的 Build、Linux UT、SCA、Coverage、两组 BVT 和 CI Required 已通过，但现有测试未覆盖上述反例。

CI：https://github.com/matrixorigin/matrixone/actions/runs/35249685469

File: pkg/sql/plan/apply_indices_fulltext.go [hunk 1]
```diff
@@ -2346,6 +2368,228 @@ func (builder *QueryBuilder) getWrappedFullTextMatches(projNode, scanNode *plan.
...
+		if fn == nil || fn.Func == nil || !safeAggForFullTextDriver(fn.Func.ObjName) || len(fn.Args) != 1 {
+			return false
+		}
+		m := monotoneWrappedFullTextMatch(fn.Args[0])
+		if m == nil {
+			return false
+		}
+		mfn := m.GetF()
+		ok := false
+		for _, d := range drivers {
```
Comment (aptend) [inline pkg/sql/plan/apply_indices_fulltext.go:2435]: [P1] Reject distinct aggregate MATCH drivers here (or preserve their per-group semantics) instead of accepting each aggregate when it matches any driver. applyJoinFullTextIndices combines multiple streams with INNER JOIN on doc_id before AGG, which proves that one document matches every pattern; HAVING MAX(match(alpha)) > 0 AND MAX(match(beta)) > 0 only proves that each pattern occurs somewhere in the group. Counterexample: cat=10 has row 1 body='alpha' and row 2 body='beta'. SELECT cat, MAX(MATCH(body) AGAINST('alpha')) a, MAX(MATCH(body) AGAINST('beta')) b FROM docs GROUP BY cat HAVING a > 0 AND b > 0 must return cat 10, but the generated alpha/beta doc-id intersection is empty and silently drops the group. A temporary typed-plan test confirmed getFullTextMatchFromAggHaving returns both distinct drivers. This affects classic FULLTEXT and FULLTEXT2.

File: pkg/sql/plan/apply_indices.go [hunk 2]
```diff
@@ -854,12 +855,19 @@ func (builder *QueryBuilder) resolveFullTextIndexPath(projNode *plan.Node) *full
 		}
 	}
 
+	var havingNode *plan.Node
 	for node := projNode; node != nil && len(node.Children) == 1; node = builder.qry.Nodes[node.Children[0]] {
+		if node.NodeType == plan.Node_FILTER {
```
Comment (aptend) [inline pkg/sql/plan/apply_indices.go:860]: [P1] This cannot classify every FILTER encountered on the single-child walk as HAVING; the walk crosses WINDOW (and can cross other cardinality/order-sensitive boundaries). Concrete query: SELECT cat, score, rn FROM (SELECT cat, MAX(MATCH(body) AGAINST('alpha')) score, ROW_NUMBER() OVER (ORDER BY cat) rn FROM docs GROUP BY cat) q WHERE score > 0. With groups cat1=nonmatch, cat2=alpha, cat3=alpha, correct output is (cat2,rn=2),(cat3,rn=3), because ROW_NUMBER is computed before the outer filter. This code harvests that outer FILTER, drives the index below AGG, removes cat1 before WINDOW, and silently returns rn=1,2. A temporary PROJECT->FILTER->WINDOW->AGG->SCAN plan test confirmed applyIndices inserts the fulltext scan. Please identify the FILTER that semantically belongs to this AGG (or stop at WINDOW/LIMIT/FILL and similar barriers) rather than using an arbitrary ancestor.

**Review by aptend — CHANGES_REQUESTED** [review]
Deep review at exact head aeef9dab7a5c916aacb7fdcc6665de219a03c056. REQUEST_CHANGES: I found two P1 silent-wrong-result paths. The single-driver MAX/SUM idea is sound, but the current invariance gate admits distinct MATCH drivers that are intersected per document, and the path resolver can treat a post-WINDOW filter as HAVING and move its membership effect below the window. Temporary typed-plan counterexamples failed as expected: the first collector returned two drivers, and the second inserted one fulltext scan beneath AGG for a FILTER above WINDOW. The PR's targeted helper/aggregate-HAVING tests pass, and exact-head required CI is green; the added BVTs do not cover these multi-row group and semantic-boundary cases.

**Review by aunjgr — CHANGES_REQUESTED** [review]
[P1] Multiple distinct aggregate MATCH drivers are not result-preserving with this rewrite. aggOutputInvariantToMatcherFilter only verifies that each aggregate MATCH appears somewhere in the candidate set. applyJoinFullTextIndices then INNER JOINs every candidate's index stream by document ID before grouping.\n\nA group can have one row matching 'alpha' and another row matching 'beta'. It correctly satisfies:\n\nHAVING MAX(MATCH(body) AGAINST('alpha')) > 0\n   AND MAX(MATCH(body) AGAINST('beta')) > 0\n\nAfter this rewrite, a single document must match both searches, so the group disappears and the query silently returns the wrong result.\n\nThe safe immediate fix is to permit aggregate-HAVING driving only when every collected candidate is equivalent to one unique MATCH expression. Supporting distinct MATCH aggregates requires group-level existential semantics rather than document-level stream intersection. Please add classic fulltext and fulltext2 regressions where separate rows in one group satisfy the two searches.

**Comment by cpegeric** [comment]
fixed

**Review by XuPeng-SH — CHANGES_REQUESTED** [review]
P1 blocker at pkg/sql/plan/apply_indices_fulltext.go:2534-2543: the membership proof unwraps the monotone scalar and reads the wrapper input instead of the actual comparison threshold. For example, HAVING score >= FLOOR(1e-1) has an effective threshold of 0, so zero-score groups must remain eligible; the current proof treats 0.1 as positive and inserts an aggregate-preceding INNER JOIN, dropping non-matching groups. With the docs fixture this loses ids 3 and 5. The query must be rejected for this optimization or the full constant expression must be evaluated before proving a positive threshold. Please add a regression for wrapped constant thresholds; FULLTEXT and FULLTEXT2 share this path. Current head verified: ef645586aef3603f12cc8aa25cc65f013b73bd2a.

**Review by XuPeng-SH — CHANGES_REQUESTED** [review]
Astra medium review on head 50f175956a26176ea8728477df403aa5f93d114b.\n\n发现 1 个 P1 聚合结果正确性问题：\n\n[P1] 包装表达式的 NULL 语义使聚合前过滤改变结果。位置：pkg/sql/plan/apply_indices_fulltext.go:2429。\n\naggOutputInvariantToMatcherFilter 通过 monotoneWrappedFullTextMatch 剥掉包装，只检查内部 MATCH 是否与 driver 相同，却没有证明包装表达式的其他参数及 NULL 行为在删除非匹配行后保持不变。ROUND(score, digits) 支持逐行精度，digits=NULL 时返回 NULL。\n\n可复现场景：同一 cat 分组中，一行 body='alpha', digits=NULL，另一行 body='beta', digits=0；查询 SUM(ROUND(MATCH(body) AGAINST('alpha'), digits))，并以 HAVING MAX(MATCH(body) AGAINST('alpha')) > 0 驱动 fulltext index。原始聚合输入为 [NULL, 0]，结果应为 0；当前 INNER JOIN 只保留匹配行后输入变为 [NULL]，结果变为 NULL。若再过滤 total=0，会直接错误丢失整个结果组。FULLTEXT2 共用同一证明和改写路径，也受影响。\n\n建议把“找到内部 MATCH”与“删除非匹配行保持完整聚合结果”拆成独立证明；保守方案先拒绝逐行、可空参数的包装表达式，并补充 FULLTEXT/FULLTEXT2 多行组混合匹配/非匹配行的数值与 NULL 回归测试。\n\n旧的 FLOOR 常量折叠问题已修复；未发现其他 P0/P1/P2 问题。

File: pkg/sql/plan/apply_indices_fulltext.go [hunk 3]
```diff
@@ -2346,6 +2369,261 @@ func (builder *QueryBuilder) getWrappedFullTextMatches(projNode, scanNode *plan.
...
+func (builder *QueryBuilder) aggOutputInvariantToMatcherFilter(aggNode *plan.Node, drivers []*plan.Expr) bool {
+	if aggNode == nil || len(aggNode.AggList) == 0 {
+		return false
+	}
+	for _, agg := range aggNode.AggList {
+		fn := agg.GetF()
+		if fn == nil || fn.Func == nil || !safeAggForFullTextDriver(fn.Func.ObjName) || len(fn.Args) != 1 {
+			return false
+		}
+		m := monotoneWrappedFullTextMatch(fn.Args[0])
```
Comment (aptend) [inline pkg/sql/plan/apply_indices_fulltext.go]: Blocking correctness issue: this proof treats every ROUND/CAST/FLOOR/CEIL wrapper as invariant when nonmatching rows are removed, because monotoneWrappedFullTextMatch follows only argument 0 and ignores wrapper nullability and additional per-row arguments. Counterexample: a group has (body=alpha,digits=NULL) and (body=beta,digits=0), and the query computes SUM(ROUND(MATCH(body) AGAINST (alpha), digits)) AS total plus MAX(MATCH(...)) AS score, with HAVING score > 0 AND total = 0. Before the rewrite the wrapper values are [NULL,0], SUM is 0, and the group survives. The fulltext INNER JOIN drops the beta row before aggregation, leaving [NULL], so SUM is NULL and the group is silently removed. This affects both fulltext implementations. A temporary typed-plan regression reproduced the bad proof: getFullTextMatchFromAggHaving returned one driver instead of none. Please separate nested MATCH discovery from a proof that the complete aggregate argument maps every excluded nonmatch row to a non-NULL zero; conservatively accept only bare MATCH or wrappers with proven zero/null behavior and constant, non-null extra arguments, and add classic/fulltext2 BVT coverage for a mixed group with nullable digits.

**Review by aptend — CHANGES_REQUESTED** [review]
Deep re-review of exact head 50f175956a26176ea8728477df403aa5f93d114b. The earlier distinct-MATCH and filter-above-window blockers are fixed, including the wrapped constant-threshold case. One aggregate-preservation blocker remains in the inline comment. Validation: the focused existing planner tests pass; a temporary typed-plan counterexample fails because the helper returns one fulltext driver when it must conservatively return none.

**Review by aunjgr — APPROVED** [review]
Reviewed 50f175956a26176ea8728477df403aa5f93d114b. My previous distinct-MATCH blocker is fixed: aggregate-HAVING driving now requires equivalent MATCH candidates, avoiding document-level intersection for group-level existential predicates. I also checked the new window/LIMIT barriers, full constant-threshold evaluation, co-aggregate rejection, and classic/fulltext2 regression coverage. No remaining blocker found. CI is green on this revision.

**Review by XuPeng-SH — CHANGES_REQUESTED** [review]
GPT-6 high review — REQUEST_CHANGES

发现 1 个 P1：fulltext index eligibility 对 CAST 链无条件放行，会导致静默错误结果。

位置：
- [apply_indices_fulltext.go:2090](https://github.com/matrixorigin/matrixone/blob/a545ec29b67ffb6d512d934d143f8b76fe76ccf7/pkg/sql/plan/apply_indices_fulltext.go#L2090)：聚合不变性检查默认所有 CAST 都保持零值；
- [apply_indices_fulltext.go:2035](https://github.com/matrixorigin/matrixone/blob/a545ec29b67ffb6d512d934d143f8b76fe76ccf7/pkg/sql/plan/apply_indices_fulltext.go#L2035)：HAVING membership 判断直接剥离 CAST。

真实 SQL 中 `CAST(CAST(0 AS CHAR) AS YEAR)` 会得到 2000，因此不能把完整 CAST 包装链当作对零值保持不变。

最小复现：四行数据 `(10,'alpha'),(10,'beta'),(20,'gamma'),(30,'delta')`，按 cat 分组。
- `SUM(CAST(CAST(MATCH(...) AS CHAR) AS YEAR)))，HAVING 使用原始 `MAX(score)>0`：正确结果为 `(10,2000)`，当前 head 返回 `(10,NULL)`；
- HAVING 将 `MAX(MATCH(...))` 依次 CAST 为 SIGNED、CHAR、YEAR 后判断 `>0`：正确应包含 cat 10、20、30，当前仅包含 cat 10。

FULLTEXT 与 FULLTEXT2 均可复现；独立对照、prepared 重复执行和另一 CN 入口也复现了错误。请对 CAST 链逐层证明零值/NULL 语义，无法证明时保守拒绝该 fulltext index 优化。此前不同 MATCH 交集、WINDOW/LIMIT、FLOOR 和 ROUND 的历史问题已闭合。

当前 CI 主检查通过，但上述 focused planner 之外的 teardown 曾因 ISCP readiness timeout 退出码为 1；这不是本次主要 blocker。

**Comment by cpegeric** [comment]
Thanks — confirmed the P1 and fixed it. Summary, and since it's the key design decision, why the outcome is a refusal (20105) 
  rather than a corrected result.

  Root cause. wrappedMatchDropSafe, unwrapMonotoneScalar, and monotoneWrappedFullTextMatch saw through any CAST unconditionally,
  assuming a cast maps a zero relevance to zero. That's false: CAST(0 AS YEAR)=2000 and CAST(CAST(0 AS CHAR) AS YEAR)=2000, so a
  dropped non-matching row (relevance 0) becomes a non-zero value. Added castPreservesZero (true only for plain numeric target
  types — int/uint/float/decimal) and gated the cast branch of all three helpers on it.

  Why reject, not solve. Driving the index INNER-joins the matchers before aggregation, physically dropping each group's
  non-matching rows; that rewrite is legal only when the result is provably invariant to dropping them. When it isn't, this
  feature uniformly refuses with 20105 — MIN, a co-aggregate like COUNT(*), per-row round digits, multiple distinct matches. The
  cast guard just adds non-zero-preserving casts to that same set.

  I'm not claiming the correct answer is uncomputable — an outer/LEFT-join scored plan that keeps non-matchers at relevance 0
  could produce it. The choice to refuse is deliberate scope and consistency:
  - This PR is a P1 correctness fix; its job is to stop silently-wrong results ((10,NULL) / partial groups), which reject does,
  minimally and safely.
  - Refusal is the established, uniform contract for every not-drop-safe aggregate-MATCH shape. Making the cast case alone fall
  back to an outer-scored plan would be an inconsistent special-case and a behavior expansion with its own correctness surface
  (per-row zero-fill for non-matchers, its interaction with the membership HAVING, etc.) — that belongs in a separate
  enhancement, evaluated for the whole family, not smuggled into this fix. 
  - It matches the reviewer's guidance to conservatively reject when zero/NULL semantics can't be proven layer by layer.

  Not over-rejected. A zero-preserving numeric cast still drives correctly — e.g. SUM(CAST(match AS DECIMAL(10,2))) with a
  membership HAVING keeps its fulltext_index_scan plan and returns the correct value; only casts through
  YEAR/CHAR/temporal/bool/unknown types are refused.
  
  Coverage. Unit tests (TestWrappedMatchDropSafe, TestUnwrapMonotoneScalar, new TestMonotoneWrappedFullTextMatchCastGuard) fail
  on the old unconditional-cast behavior and pass on the fix. BVT added to fulltext_agg_having and fulltext2_agg_having (the
  exact 4-row cat grouping): the two cast-chain queries assert 20105, plus a driven numeric-cast control; both pass genrs + run
  mode (-g -n) at 100%. Prior closures (mixed-MATCH intersection, WINDOW/LIMIT, FLOOR/ROUND, per-row round digits) are untouched.

**Review by XuPeng-SH — CHANGES_REQUESTED** [review]
GPT-6 High deep review of exact head 6a96f496e4e773b83e201e9d3714c769409aada4: REQUEST_CHANGES.

P2 — The FULLTEXT2 numeric-cast positive control is not synchronized and its expected result accepts no rows. castrd is created and populated, but unlike the preceding FULLTEXT2 fixtures it is not force-reindexed before the queries. The intended query should preserve cat 10; however, the result currently ends without asserting a row. If async indexing has not caught up, the query may return empty and the regression passes without proving aggregate preservation; if indexing does catch up, the expected result may fail. Add force_sync for castft, assert the nonempty expected row(s), and regenerate/verify by normal comparison. [SQL setup/query](https://github.com/matrixorigin/matrixone/blob/6a96f496e4e773b83e201e9d3714c769409aada4/test/distributed/cases/fulltext2/fulltext2_agg_having.sql#L121) · [current result](https://github.com/matrixorigin/matrixone/blob/6a96f496e4e773b83e201e9d3714c769409aada4/test/distributed/cases/fulltext2/fulltext2_agg_having.result#L154).

The previous nested-cast safety blocker is addressed. Review also verified the distinct-MATCH, co-aggregate, nullable ROUND, folded-threshold, and window-boundary protections. Exact-head CI passed, but does not establish this positive control has a nonempty oracle.

**Review by XuPeng-SH — CHANGES_REQUESTED** [review]
GPT-6 High deep re-review of exact head `a348bf31a188e0d665dd9ff8bcf1c83332a05db4`: REQUEST_CHANGES.

**[P1] OFFSET without LIMIT still allows aggregate-match filtering before pagination.**

The new barrier in `pkg/sql/plan/apply_indices.go:isFullTextAggHavingBarrier` checks only `node.Limit != nil`; it does not treat a non-zero `node.Offset` as a barrier. The earlier predicate pushdown in `pkg/sql/plan/pushdown.go:298` likewise protects only LIMIT. Thus an outer predicate can be pushed through an OFFSET-only sort into `AGG.FilterList`, and `getFullTextMatchFromAggHaving` then converts it into an index driver, filtering groups before OFFSET is applied.

A reachable example is:

```sql
select cat from (
  select cat, max(match(body) against('alpha')) as score
  from docs
  group by cat
  order by cat offset 1
) q
where score > 0
order by cat;
```

With ordered groups `(1,'nomatch'), (2,'alpha'), (3,'alpha')`, the correct result is cats 2 and 3. Driving the index first removes cat 1, then OFFSET skips cat 2, returning only cat 3. Please preserve OFFSET as a semantic barrier in both predicate pushdown and aggregate-HAVING driver resolution, and add FULLTEXT/FULLTEXT2 SQL regressions with OFFSET but no LIMIT.

The earlier distinct-MATCH, WINDOW/LIMIT, threshold, nullable ROUND, CAST-chain, and FULLTEXT2 positive-control findings appear addressed. Exact-head CI is green, including planner shards and both aggregate-HAVING BVTs (42/42 each); neither covers OFFSET-only pagination. The counterexample above is source-traced, not locally executed.

**Review by XuPeng-SH — APPROVED** [review]
GPT-6 High deep re-review — APPROVE at exact head acc5b1f6f9e7c8952f597a96fd4c6cedc1e3a007.

The prior OFFSET-without-LIMIT blocker is closed: predicate pushdown now treats non-zero OFFSET as a barrier, and aggregate-HAVING driver resolution independently rejects OFFSET nodes. The added FULLTEXT and FULLTEXT2 SQL cases cover the outer filter across OFFSET and a real HAVING/OFFSET positive control. Reviewed the remaining aggregate/drop-safety guards and their regressions; no blocking findings remain. Exact-head CI is green, including planner tests and both aggregate-HAVING BVTs.
